desc/objectives/utils.py

```diff
@@ -142,17 +143,8 @@ def factorize_linear_constraints(objective, constraint):  # noqa: C901
     xp = put(xp, unfixed_idx, Ainv_full @ b)
     xp = jnp.asarray(xp)
 
-    @jit
-    def project(x):
-        """Project a full state vector into the reduced optimization vector."""
-        x_reduced = Z.T @ ((x - xp)[unfixed_idx])
-        return jnp.atleast_1d(jnp.squeeze(x_reduced))
-
-    @jit
-    def recover(x_reduced):
-        """Recover the full state vector from the reduced optimization vector."""
-        dx = put(jnp.zeros(objective.dim_x), unfixed_idx, Z @ x_reduced)
-        return jnp.atleast_1d(jnp.squeeze(xp + dx))
+    project = _Project(Z, xp, unfixed_idx)
```

Comment:
This is just not to jit these functions until we use them, right? And maybe to save the null-space etc?

Reply:
No it's more about avoiding recompilation. Right now every time you call `factorize_linear_constraints` you get a new `project/recover` function which gets compiled each time, since its a new function each time. By making them classes with jitted methods the compile cache is shared between class instances so it will only compile once (assuming size/shape doesnt change etc)

Reply:
ah ok makes sense